File: src/judgemetrics/schemas/metrics.py
# src/judgemetrics/schemas/metrics.py
"""Metric responses: the registry, observations, subject metrics, compare rows,
the provenance chain of an observation, and the corrections intake.

Every observation leaves the API with the brief's presentation fields —
numerator, denominator, eligible count (sample size), the period (date
range), the source's coverage window and whether the outcome is observable
in it, the interval and its method where one is defined, the suppression
flag and threshold, the methodology version, and a methodology link — and
never a person: members are entity ids and the rows carry subject ids only
(``tests/golden/test_public_contract.py`` walks every metrics route for a
person key or hash). Suppression is enforced here, at the schema layer:
whenever ``suppressed`` is true the validator nulls ``numerator``,
``denominator``, ``rate``, ``value``, ``distribution``, ``lower``, and
``upper`` whatever the caller passed, so a stored number below the
threshold cannot reach a client through any construction path; the
``eligible_count`` and the ``suppression_threshold`` stay, which is how a
reader learns why the number was withheld. ``interval_method`` follows the
metric's kind: ``wilson`` for shares and fixed-window rates, ``greenwood``
for Kaplan-Meier estimates, none otherwise.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal, Self

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator, model_validator

from judgemetrics.schemas.common import Page
from judgemetrics.schemas.judges import SYNTHETIC_DESCRIPTION, CourtRef

MetricKind = Literal["count", "share", "windowed_rate", "survival", "distribution", "median"]
MetricSubjectType = Literal["judge", "court"]
IntervalMethod = Literal["wilson", "greenwood"]
CompareSort = Literal["rate", "numerator", "denominator", "value", "name"]
SortOrder = Literal["asc", "desc"]
CorrectionTargetType = Literal["judge", "court", "case", "metric_observation"]
CorrectionStatusOut = Literal["received"]
MemberKind = Literal["decision", "charge", "court_case", "sentence", "court_event", "justice_event"]

# The interval every rate of a kind publishes (docs/METHODOLOGY.md "How to read a number").
INTERVAL_METHOD_BY_KIND: dict[str, IntervalMethod] = {
    "share": "wilson",
    "windowed_rate": "wilson",
    "survival": "greenwood",
}
# The figures a suppressed observation never carries.
SUPPRESSED_FIELDS: tuple[str, ...] = (
    "numerator",
    "denominator",
    "rate",
    "value",
    "distribution",
    "lower",
    "upper",
)
METHODOLOGY_URL_DESCRIPTION = (
    "The methodology page anchored at the metric's slug: how the number is computed, "
    "its attribution rule, and the known limitations."
)
SHA256_PATTERN = r"^[0-9a-f]{64}$"


def interval_method_for(kind: str) -> IntervalMethod | None:
    return INTERVAL_METHOD_BY_KIND.get(kind)


# --- the registry ---------------------------------------------------------------------------


class AttributionOut(BaseModel):
    """A metric's structured inclusion rule (docs/METHODOLOGY.md \"Attribution\")."""

    decision_type: str | None = Field(description="The decision type a row must carry, if any.")
    actor_types: list[str] | None = Field(description="The actor types admitted, if restricted.")
    discretion: list[str] | None = Field(
        description="The discretion classifications admitted, if restricted."
    )
    assignment_gate: str = Field(
        description=(
            "How a row is tied to the subject: deciding_judge, assigned_at_time, "
            "assigned_ever, sentencing_judge, or court_of_case."
        )
    )


class MetricDefinitionOut(BaseModel):
    """One registry entry: what a number means and how it is computed."""

    slug: str
    name: str
    kind: MetricKind
    subject_types: list[MetricSubjectType]
    description: str
    numerator: str
    denominator: str
    eligibility: str
    attribution: AttributionOut
    index_event: str | None = Field(
        description="pretrial_release, disposition, or sentence for a windowed metric."
    )
    outcome: str | None = Field(description="The justice_event_type a windowed metric counts.")
    windows_days: list[int] | None = Field(description="The follow-up windows, in days.")
    dimension: str | None = Field(description="disposition or offense_category, when grouped.")
    suppression_threshold: int = Field(
        ge=0, description="The denominator below which an observation is suppressed."
    )
    unit: Literal["count", "share", "days"]
    version: str = Field(description="The definition's version; bumped when its semantics change.")
    methodology_url: str = Field(description=METHODOLOGY_URL_DESCRIPTION)


class SuppressionOut(BaseModel):
    default_threshold: int = Field(ge=0)
    rule: str
    rationale: str


class MethodologyTerm(BaseModel):
    """One term of the methodology prose: a heading and its text, rendered as text nodes."""

    term: str
    text: str


class MethodologyChange(BaseModel):
    """One methodology changelog entry."""

    version: str = Field(description="The methodology version the entry describes.")
    text: str


class Registry(BaseModel):
    """The versioned metric registry the methodology page is rendered from.

    The prose sections (``how_to_read``, ``semantics``, ``attribution_notes``,
    ``changelog``) are the same constants ``docs/METHODOLOGY.md`` is rendered
    from, so the web page and the committed document never diverge.
    """

    registry_version: int = Field(ge=1)
    methodology_version: str
    methodology_url: str = Field(description="The methodology page.")
    windows_days: list[int] = Field(
        description="The follow-up windows every windowed metric is computed over, in days."
    )
    known_limitations: list[str] = Field(
        description="The brief's statistical warnings, verbatim and never softened."
    )
    suppression: SuppressionOut
    how_to_read: list[MethodologyTerm] = Field(
        description="What each presentation field beside a number means."
    )
    semantics: list[MethodologyTerm] = Field(
        description="Index events, exposure, outcomes and windows, censoring, Kaplan-Meier."
    )
    attribution_notes: list[str] = Field(
        description="How rows are tied to a judge, and what is never attributed."
    )
    gate_descriptions: dict[str, str] = Field(
        description="Each assignment gate (`AttributionOut.assignment_gate`) in words."
    )
    changelog: list[MethodologyChange] = Field(description="Oldest version first.")
    definitions: list[MetricDefinitionOut] = Field(description="In registry order.")


# --- observations ---------------------------------------------------------------------------


class ObservationCoverage(BaseModel):
    """What the observation's source covers, and whether it can document the outcome."""

    coverage_start: date | None = Field(description="First day the source's records cover.")
    coverage_end: date | None = Field(
        description="Last day the source's records cover; follow-up is censored the day after."
    )
    observable: bool = Field(
        description=(
            "True when the source documents the metric's outcome (always true for a metric "
            "without an outcome); a metric that is not observable is never published."
        )
    )


class SuppressibleFigures(BaseModel):
    """The figures a suppressed row withholds; the validator nulls them whenever ``suppressed``."""

    numerator: int | None = Field(
        ge=0, description="The observed count: rows or members meeting the metric's condition."
    )
    denominator: int | None = Field(
        ge=0,
        description=(
            "The cohort size the numerator is divided by (the followed members of a "
            "fixed-window rate; the whole cohort of a survival estimate; the attributed rows "
            "of a share; the values a median is taken over; the population of a count)."
        ),
    )
    rate: float | None = Field(
        ge=0.0,
        le=1.0,
        description=(
            "numerator / denominator for a share or fixed-window rate; 1 - S(w) for a "
            "survival estimate; six decimals; null for counts, distributions, and medians."
        ),
    )
    value: float | None = Field(description="A median, in days; null for every other kind.")
    distribution: dict[str, int] | None = Field(
        description="The whole map of a distribution (vocabulary value → count); null otherwise."
    )
    lower: float | None = Field(ge=0.0, le=1.0, description="The 95% interval's lower bound.")
    upper: float | None = Field(ge=0.0, le=1.0, description="The 95% interval's upper bound.")
    suppressed: bool = Field(
        description=(
            "True when the denominator is below the metric's suppression threshold: the "
            "numerator, denominator, rate, value, distribution, and interval are withheld."
        )
    )
    suppression_threshold: int = Field(
        ge=0, description="The metric's threshold, stated so a reader knows why."
    )

    @model_validator(mode="after")
    def _withhold_when_suppressed(self) -> Self:
        if self.suppressed:
            for name in SUPPRESSED_FIELDS:
                setattr(self, name, None)
        return self


class Observation(SuppressibleFigures):
    """One current metric observation with every presentation field."""

    id: uuid.UUID
    slug: str
    name: str = Field(description="The metric's name from the registry.")
    kind: MetricKind
    unit: Literal["count", "share", "days"]
    version: str = Field(description="The definition version the observation was computed under.")
    subject_type: MetricSubjectType
    subject_id: uuid.UUID
    source: str = Field(description="Source register key (docs/DATA_SOURCES.md).")
    synthetic: bool = Field(description=SYNTHETIC_DESCRIPTION)
    period_start: date = Field(description="The observation's date range: the source's window.")
    period_end: date
    window_days: int | None = Field(description="The follow-up window of a windowed metric.")
    dimension_value: str | None = Field(description="The group of a dimensioned metric.")
    eligible_count: int = Field(
        ge=0,
        description=(
            "Sample size: the whole cohort before any follow-up restriction, published even "
            "when the number is suppressed."
        ),
    )
    interval_method: IntervalMethod | None = Field(
        description="wilson for shares and fixed-window rates, greenwood for survival estimates."
    )
    coverage: ObservationCoverage
    methodology_version: str
    methodology_url: str = Field(description=METHODOLOGY_URL_DESCRIPTION)
    snapshot_hash: str = Field(
        min_length=64,
        max_length=64,
        pattern=SHA256_PATTERN,
        description="The content hash of the exported tables the number was computed from.",
    )
    computed_at: datetime


class SubjectSummary(BaseModel):
    subject_type: MetricSubjectType
    id: uuid.UUID
    canonical_name: str
    synthetic: bool = Field(description=SYNTHETIC_DESCRIPTION)


class SubjectMetrics(BaseModel):
    """Every current observation of a judge or court, grouped by metric slug."""

    subject: SubjectSummary
    registry_version: int
    methodology_version: str
    methodology_url: str = Field(description="The methodology page.")
    total: int = Field(ge=0, description="Observations across every slug.")
    observations: dict[str, list[Observation]] = Field(
        description=(
            "By metric slug, each list ordered by window, dimension value, and source; a "
            "metric the source cannot observe has no entry."
        )
    )


# --- compare ---------------------------------------------------------------------------------


class CompareCohort(BaseModel):
    """What the rows are compared within."""

    metric: str = Field(description="The metric slug.")
    version: str = Field(description="The definition version compared.")
    window_days: int | None
    court_id: uuid.UUID | None
    jurisdiction_id: uuid.UUID | None
    name: str = Field(description="The court's or jurisdiction's name.")
    period_start: date | None = Field(
        description=(
            "The cohort's reference period: the requested one, else the period most rows of "
            "the whole cohort share; null when the cohort has no row."
        )
    )
    period_end: date | None
    sort: CompareSort
    order: SortOrder


class CompareRow(SuppressibleFigures):
    """One judge's observation of the compared metric."""

    subject_id: uuid.UUID
    name: str = Field(description="The judge's canonical name.")
    court: CourtRef = Field(description="The judge's court within the cohort.")
    synthetic: bool = Field(description=SYNTHETIC_DESCRIPTION)
    observation_id: uuid.UUID
    source: str
    period_start: date
    period_end: date
    window_days: int | None
    dimension_value: str | None
    eligible_count: int = Field(ge=0, description="Sample size before any follow-up restriction.")
    interval_method: IntervalMethod | None
    coverage_warning: str | None = Field(
        description=(
            "Set when the judge's coverage window differs from the cohort's reference period, "
            "or the source cannot document the metric's outcome."
        )
    )


class ComparePage(Page[CompareRow]):
    """A sorted, paginated compare table with its cohort and methodology metadata."""

    cohort: CompareCohort
    methodology_version: str
    methodology_url: str = Field(description=METHODOLOGY_URL_DESCRIPTION)


# --- provenance ---------------------------------------------------------------------------


class TracedObservation(Observation):
    """The observation at the top of a provenance chain, with its versions."""

    registry_version: int
    code_version: str = Field(description="The package version and git SHA that computed it.")
    superseded_at: datetime | None = Field(
        description="Null for a current observation (the only kind the API traces)."
    )


class SnapshotOut(BaseModel):
    """The hashed export the observation was computed from."""

    content_hash: str = Field(min_length=64, max_length=64, pattern=SHA256_PATTERN)
    label: str | None
    exported_at: datetime
    code_version: str
    registry_version: int
    methodology_version: str
    row_counts: dict[str, int] = Field(description="Rows per exported table.")


class MemberGroup(BaseModel):
    """The observation's members of one kind: the eligible canonical rows behind the number."""

    member_kind: MemberKind
    members: int = Field(ge=0, description="Rows of this kind behind the observation.")
    counted: int = Field(ge=0, description="Members in the numerator.")
    followed: int = Field(ge=0, description="Members in the denominator after censoring.")
    resolved: int = Field(
        ge=0,
        description="Members whose canonical row still exists; equals `members` when complete.",
    )
    case_ids: list[uuid.UUID] = Field(
        description="The distinct cases the members belong to (`/cases/{id}`), sorted."
    )


class SourceRecordOut(BaseModel):
    """One retrieved raw artifact behind the members."""

    id: uuid.UUID
    source: str = Field(description="Source register key.")
    external_record_id: str | None = Field(description="The artifact's id at the source.")
    raw_sha256: str = Field(min_length=64, max_length=64, pattern=SHA256_PATTERN)
    retrieved_at: datetime
    parser_version: str
    ingest_run_id: uuid.UUID
    artifact_uri: str | None = Field(
        description=(
            "Where the artifact was fetched from, when that is a public http(s) URL; null "
            "for an artifact read from the operator's filesystem (a fixture or the synthetic "
            "dataset)."
        )
    )


class SourceOut(BaseModel):
    """The source system a record came from."""

    source: str = Field(description="Source register key.")
    owner: str
    source_type: str
    synthetic: bool = Field(description=SYNTHETIC_DESCRIPTION)
    coverage_start: date | None
    coverage_end: date | None
    observable_outcomes: list[str]


class ObservationProvenance(BaseModel):
    """The brief's chain, top-down: observation → snapshot → members → cases → records → sources."""

    observation: TracedObservation
    snapshot: SnapshotOut
    members: list[MemberGroup] = Field(description="By member kind.")
    source_records: list[SourceRecordOut] = Field(
        description="The distinct artifacts behind every member, newest retrieval first."
    )
    sources: list[SourceOut] = Field(description="The distinct source systems, by key.")
    complete: bool = Field(
        description=(
            "True when every member resolved to a canonical row and every row to a source "
            "record with its artifact digest."
        )
    )


# --- corrections ----------------------------------------------------------------------------


class CorrectionIn(BaseModel):
    """A public data-correction request. The contact is encrypted before it is stored."""

    model_config = ConfigDict(str_strip_whitespace=True, extra="forbid")

    target_type: CorrectionTargetType
    target_id: uuid.UUID
    reason: str = Field(
        min_length=20, max_length=4000, description="What is wrong and what the record should say."
    )
    contact: str = Field(
        min_length=3,
        max_length=320,
        description="How to reach the requester about this request; stored encrypted.",
    )
    supporting_material: HttpUrl | None = Field(
        default=None, description="An optional http(s) link to supporting material."
    )

    @field_validator("supporting_material")
    @classmethod
    def _bounded_url(cls, value: HttpUrl | None) -> HttpUrl | None:
        if value is not None and len(str(value)) > 2000:
            msg = "must be at most 2000 characters"
            raise ValueError(msg)
        return value

    @field_validator("reason", "contact")
    @classmethod
    def _printable(cls, value: str) -> str:
        if any(ch.isspace() and ch not in " \n\r\t" for ch in value) or any(
            ord(ch) < 32 and ch not in "\n\r\t" for ch in value
        ):
            msg = "must not contain control characters"
            raise ValueError(msg)
        return value


class CorrectionAccepted(BaseModel):
    """The acknowledgement: the request's id and status, nothing the requester submitted."""

    id: uuid.UUID
    status: CorrectionStatusOut
    received_at: datetime


__all__ = [
    "INTERVAL_METHOD_BY_KIND",
    "SUPPRESSED_FIELDS",
    "AttributionOut",
    "CompareCohort",
    "ComparePage",
    "CompareRow",
    "CompareSort",
    "CorrectionAccepted",
    "CorrectionIn",
    "CorrectionTargetType",
    "IntervalMethod",
    "MemberGroup",
    "MethodologyChange",
    "MethodologyTerm",
    "MetricDefinitionOut",
    "MetricKind",
    "MetricSubjectType",
    "Observation",
    "ObservationCoverage",
    "ObservationProvenance",
    "Registry",
    "SnapshotOut",
    "SortOrder",
    "SourceOut",
    "SourceRecordOut",
    "SubjectMetrics",
    "SubjectSummary",
    "SuppressibleFigures",
    "SuppressionOut",
    "TracedObservation",
    "interval_method_for",
]

File: web/lib/metrics.ts
// web/lib/metrics.ts
// Helpers for the metric surfaces: grouping a subject's observations by
// slug → window → dimension, choosing a window, the formatting every
// presentation field uses (rates as percentages, medians as days, counts as
// integers, periods and coverage as date ranges), suppression, the cohort
// labels, and the methodology anchor. Every formatter is null-safe so a
// suppressed or not-observable row formats as a dash, never as "NaN%". The
// follow-up windows are read from the registry response (`windows_days`),
// never hard-coded here.
import type {
  CompareRow,
  MetricDefinition,
  Observation,
  Registry,
  Schemas,
} from "@/lib/api/client";
import { formatDate, formatInteger, titleCase } from "@/lib/format";

export type ObservationCoverage = Schemas["ObservationCoverage"];

/**
 * The brief's statement, word for word, above every metric panel and on the
 * methodology page (`data-testid="association-statement"` on both).
 */
export const ASSOCIATION_STATEMENT =
  "These statistics describe associations in available records. They do not prove that a judicial decision caused a later event.";

/** The window key of an unwindowed metric. */
export const NO_WINDOW = "all";
/** The dimension key of an undimensioned observation. */
export const NO_DIMENSION = "";
/** The window the judge page and the compare page open on. */
export const DEFAULT_WINDOW_DAYS = 365;

/** Dimension key → the observations of that dimension, one per source, in API order. */
export type DimensionGroup = Record<string, Observation[]>;
/** Window key → dimension group. */
export type WindowGroup = Record<string, DimensionGroup>;
/** Metric slug → window group. */
export type GroupedObservations = Record<string, WindowGroup>;

export function windowKey(days: number | null | undefined): string {
  return days === null || days === undefined ? NO_WINDOW : String(days);
}

export function dimensionKey(value: string | null | undefined): string {
  return value ?? NO_DIMENSION;
}

/**
 * `SubjectMetrics.observations` (a map keyed by slug, each list ordered by
 * window, dimension value, and source) grouped by slug → window → dimension.
 */
export function groupObservations(
  observations: Record<string, Observation[]>,
): GroupedObservations {
  const grouped: GroupedObservations = {};
  for (const [slug, rows] of Object.entries(observations)) {
    const windows: WindowGroup = {};
    for (const row of rows) {
      const window = (windows[windowKey(row.window_days)] ??= {});
      (window[dimensionKey(row.dimension_value)] ??= []).push(row);
    }
    grouped[slug] = windows;
  }
  return grouped;
}

/** The dimension group of one window of a slug's group; `null` days is the unwindowed entry. */
export function pickWindow(
  group: WindowGroup | undefined,
  days: number | null,
): DimensionGroup | undefined {
  return group?.[windowKey(days)];
}

/** The first observation (the first source) of a slug at a window and dimension. */
export function pickObservation(
  grouped: GroupedObservations,
  slug: string,
  options: { window?: number | null; dimension?: string | null } = {},
): Observation | undefined {
  return pickWindow(grouped[slug], options.window ?? null)?.[dimensionKey(options.dimension)]?.[0];
}

/** The dimension values present for a slug at a window, in API order. */
export function dimensionValues(
  grouped: GroupedObservations,
  slug: string,
  window: number | null = null,
): string[] {
  return Object.keys(pickWindow(grouped[slug], window) ?? {}).filter((key) => key !== NO_DIMENSION);
}

/** The follow-up windows the registry declares, ascending; never a literal here. */
export function windowsFromRegistry(registry: Pick<Registry, "windows_days" | "definitions">): number[] {
  const declared = new Set<number>(registry.windows_days);
  for (const definition of registry.definitions) {
    for (const days of definition.windows_days ?? []) declared.add(days);
  }
  return [...declared].sort((a, b) => a - b);
}

/**
 * The window a page shows: the query value when it is one of the registry's
 * windows, else the default (365) when the registry has it, else the largest
 * window. `null` when the registry declares no window at all.
 */
export function resolveWindow(raw: string | undefined, windows: number[]): number | null {
  if (windows.length === 0) return null;
  const requested = Number.parseInt(raw ?? "", 10);
  if (Number.isFinite(requested) && windows.includes(requested)) return requested;
  if (windows.includes(DEFAULT_WINDOW_DAYS)) return DEFAULT_WINDOW_DAYS;
  return windows[windows.length - 1];
}

export function isSuppressed(observation: Pick<Observation, "suppressed">): boolean {
  return observation.suppressed;
}

/** "22.1%" for 0.220641; a dash for null (suppressed or absent). */
export function formatRate(rate: number | null | undefined, digits = 1): string {
  if (rate === null || rate === undefined || !Number.isFinite(rate)) return "—";
  return `${(rate * 100).toFixed(digits)}%`;
}

const INTERVAL_METHOD_LABEL: Record<string, string> = {
  wilson: "95% Wilson interval",
  greenwood: "95% Greenwood interval",
};

/** "17.6%–27.3% (95% Wilson interval)"; a dash when either bound is null. */
export function formatInterval(
  lower: number | null | undefined,
  upper: number | null | undefined,
  method: string | null | undefined,
): string {
  if (lower === null || lower === undefined || upper === null || upper === undefined) return "—";
  const label = method ? INTERVAL_METHOD_LABEL[method] ?? `95% ${titleCase(method)} interval` : "95% interval";
  return `${formatRate(lower)}–${formatRate(upper)} (${label})`;
}

/** "150.5 days" for a median; whole numbers without a decimal; a dash for null. */
export function formatDays(value: number | null | undefined): string {
  if (value === null || value === undefined || !Number.isFinite(value)) return "—";
  const rounded = Number.isInteger(value) ? String(value) : value.toFixed(1);
  return `${rounded} day${value === 1 ? "" : "s"}`;
}

/** "1,234" for a count; a dash for null. */
export function formatCount(value: number | null | undefined): string {
  if (value === null || value === undefined || !Number.isFinite(value)) return "—";
  return formatInteger(value);
}

/** "Jan 1, 2016 – Dec 31, 2023"; either end may be null. */
export function formatPeriod(start: string | null | undefined, end: string | null | undefined): string {
  if (!start && !end) return "—";
  return `${formatDate(start)} – ${formatDate(end)}`;
}

/** "synthetic · Jan 1, 2016 – Dec 31, 2023 · outcome observable". */
export function formatCoverage(coverage: ObservationCoverage, source: string): string {
  const window = formatPeriod(coverage.coverage_start, coverage.coverage_end);
  const observable = coverage.observable ? "outcome observable" : "outcome not observable";
  return `${source} · ${window} · ${observable}`;
}

/** "n / d" with both formatted; a dash when either is withheld. */
export function formatFraction(
  numerator: number | null | undefined,
  denominator: number | null | undefined,
): string {
  if (numerator === null || numerator === undefined || denominator === null || denominator === undefined) {
    return "—";
  }
  return `${formatCount(numerator)} / ${formatCount(denominator)}`;
}

/**
 * The primary figure of an observation by kind: a percentage for a share,
 * fixed-window rate, survival estimate, or one value of a distribution; days
 * for a median; an integer for a count. A dash when the figure is withheld.
 */
export function primaryFigure(
  observation: Pick<Observation, "kind" | "numerator" | "denominator" | "rate" | "value">,
): string {
  switch (observation.kind) {
    case "count":
      return formatCount(observation.numerator);
    case "median":
      return formatDays(observation.value);
    case "distribution":
      return observation.numerator !== null && observation.denominator
        ? formatRate(observation.numerator / observation.denominator)
        : "—";
    default:
      return formatRate(observation.rate);
  }
}

/** "released" for a slug like "pretrial_released"; a dimension value in words. */
export function dimensionLabel(value: string | null | undefined): string {
  return value ? titleCase(value) : "";
}

export const COHORTS = {
  court: "Same court, same period",
  jurisdiction: "Jurisdiction, same period",
} as const;

export type CohortKind = keyof typeof COHORTS;

export function isCohortKind(value: string): value is CohortKind {
  return value in COHORTS;
}

/** The cohort query value, defaulting to the court. */
export function parseCohort(raw: string | undefined): CohortKind {
  return raw && isCohortKind(raw) ? raw : "court";
}

/** "Same court, same period (Jan 1, 2016 – Dec 31, 2023)". */
export function cohortLabel(
  observation: Pick<Observation, "period_start" | "period_end">,
  kind: CohortKind = "court",
): string {
  return `${COHORTS[kind]} (${formatPeriod(observation.period_start, observation.period_end)})`;
}

/**
 * The methodology page anchored at the metric's slug. The API already sends
 * `methodology_url` in this shape; pages pass it through when present and
 * fall back to this for a definition without one. The version is carried
 * in the link title, not the path: one page serves the current registry.
 */
export function methodologyHref(slug: string, version?: string, base = "/methodology"): string {
  void version;
  return `${base}#${encodeURIComponent(slug)}`;
}

/** The definitions a subject type can carry, in registry order. */
export function definitionsFor(
  registry: Pick<Registry, "definitions">,
  subject: "judge" | "court",
): MetricDefinition[] {
  return registry.definitions.filter((definition) => definition.subject_types.includes(subject));
}

/** True for a fixed-window rate or a survival estimate. */
export function isWindowed(definition: Pick<MetricDefinition, "windows_days">): boolean {
  return (definition.windows_days?.length ?? 0) > 0;
}

/** The compare sort keys the API accepts. */
export const COMPARE_SORTS = ["rate", "numerator", "denominator", "value", "name"] as const;
export type CompareSort = (typeof COMPARE_SORTS)[number];
export const COMPARE_ORDERS = ["asc", "desc"] as const;
export type CompareOrder = (typeof COMPARE_ORDERS)[number];

/** The sort a metric's figure lives under: medians sort by value, everything else by rate. */
export function defaultSort(definition: Pick<MetricDefinition, "kind">): CompareSort {
  if (definition.kind === "median") return "value";
  if (definition.kind === "count" || definition.kind === "distribution") return "numerator";
  return "rate";
}

/** The comparable figure of a compare row: the rate, or the median's value. */
export function compareFigure(row: Pick<CompareRow, "rate" | "value">): number | null {
  return row.rate ?? row.value ?? null;
}

export interface CohortPosition {
  /** Judges in the cohort (the page total, suppressed rows included). */
  judges: number;
  /** Judges whose figure is published (unsuppressed) on the page. */
  published: number;
  /** The median of the published figures; null when none is published. */
  median: number | null;
  /** The judge's 1-based rank among the published figures, descending; null when unranked. */
  rank: number | null;
}

/**
 * Where a judge falls in a cohort's distribution of one metric, from the
 * compare rows of the same window and dimension: the number of judges, the
 * cohort's median figure, and the judge's rank by figure (highest first).
 */
export function cohortPosition(
  rows: CompareRow[],
  judgeId: string,
  total: number,
  dimension: string | null = null,
): CohortPosition {
  const cohort = rows.filter((row) => dimensionKey(row.dimension_value) === dimensionKey(dimension));
  const published = cohort
    .map((row) => ({ id: row.subject_id, figure: compareFigure(row) }))
    .filter((row): row is { id: string; figure: number } => row.figure !== null)
    .sort((a, b) => b.figure - a.figure);
  const figures = published.map((row) => row.figure);
  const mid = Math.floor(figures.length / 2);
  const median =
    figures.length === 0
      ? null
      : figures.length % 2 === 1
        ? figures[mid]
        : (figures[mid - 1] + figures[mid]) / 2;
  const index = published.findIndex((row) => row.id === judgeId);
  return {
    judges: dimension === null ? total : cohort.length,
    published: published.length,
    median,
    rank: index === -1 ? null : index + 1,
  };
}

export type IndexEvent = "pretrial_release" | "disposition" | "sentence";

export interface JudgePanelSpec {
  /** The section anchor. */
  id: "cases" | "pretrial" | "outcomes" | "disposition" | "sentencing";
  title: string;
  /** Unwindowed metrics shown in this order (registry names are the labels). */
  slugs: readonly string[];
  /** The index event whose windowed metrics (rates and survival estimates) the panel also shows. */
  indexEvent: IndexEvent | null;
  /** The cases-route filter the "View eligible cases" link carries, when one applies. */
  casesFilter: Record<string, string> | null;
}

/**
 * The judge page's panels. Windowed metrics are placed by the registry's
 * `index_event`, not by slug, so a new outcome metric lands in the right
 * panel without a change here; a judge metric no panel names is listed
 * under "Other metrics" rather than dropped.
 */
export const JUDGE_PANELS: readonly JudgePanelSpec[] = [
  {
    id: "cases",
    title: "Cases",
    slugs: ["eligible_cases", "eligible_defendants"],
    indexEvent: null,
    casesFilter: null,
  },
  {
    id: "pretrial",
    title: "Pretrial",
    slugs: ["pretrial_decisions", "pretrial_released", "pretrial_detained", "pretrial_release_share"],
    indexEvent: null,
    casesFilter: null,
  },
  {
    id: "outcomes",
    title: "Outcomes after qualifying release",
    slugs: [],
    indexEvent: "pretrial_release",
    casesFilter: null,
  },
  {
    id: "disposition",
    title: "Disposition",
    slugs: ["disposition_distribution", "judicial_dismissal_rate", "median_days_to_disposition"],
    indexEvent: "disposition",
    casesFilter: { status: "closed" },
  },
  {
    id: "sentencing",
    title: "Sentencing",
    slugs: [
      "sentence_count",
      "incarceration_days_median",
      "probation_days_median",
      "incarceration_days_median_by_offense_category",
    ],
    indexEvent: "sentence",
    casesFilter: { status: "closed" },
  },
];

/** The kinds a cohort comparison (`/metrics/compare`) is fetched for on the judge page. */
export const COMPARED_KINDS: ReadonlySet<MetricDefinition["kind"]> = new Set([
  "share",
  "windowed_rate",
  "survival",
  "median",
]);

/** The judge's case list with a panel's filter, when the cases route supports one. */
export function eligibleCasesHref(judgeId: string, filter: Record<string, string> | null): string {
  const query = filter ? new URLSearchParams(filter).toString() : "";
  return `/judges/${judgeId}/cases${query ? `?${query}` : ""}`;
}

/** The compare page for one metric, window, and cohort. */
export function compareHref(params: {
  metric: string;
  window?: number | null;
  court_id?: string;
  jurisdiction_id?: string;
  sort?: CompareSort;
  order?: CompareOrder;
  offset?: number;
}): string {
  const query = new URLSearchParams();
  query.set("metric", params.metric);
  if (params.window !== null && params.window !== undefined) query.set("window", String(params.window));
  if (params.court_id) query.set("court_id", params.court_id);
  else if (params.jurisdiction_id) query.set("jurisdiction_id", params.jurisdiction_id);
  if (params.sort) query.set("sort", params.sort);
  if (params.order) query.set("order", params.order);
  if (params.offset) query.set("offset", String(params.offset));
  return `/compare?${query.toString()}`;
}
